measurement/plugins/youtube/measurements.py

```diff
@@ -0,0 +1,129 @@
+import socket
+import urllib
+import time
+import tempfile
+import os
+
+import youtube_dl
+import validators
+from validators import ValidationFailure
+
+from measurement.measurements import BaseMeasurement
+from measurement.results import Error
+from measurement.units import RatioUnit, TimeUnit, StorageUnit, NetworkUnit
+from measurement.plugins.youtube.results import YouTubeMeasurementResult
+
+YOUTUBE_ERRORS = {
+    "youtube-download": "Download utility could not download file",
+    "youtube-extractor": "Unable to extract info from youtube",
+    "youtube-url": "Could not recognise URL",
+    "youtube-attribute": "Could not parse attributes from progress dict",
+    "youtube-progress_length": "Recorded progress dicts were too short",
+    "youtube-file": "Could not remove file!!",
+    "youtube-no_directory": "Could not find directory!!",
+    "youtube-directory_nonempty": "Could not remove directory, non-empty!",
+}
+
+
+class YouTubeMeasurement(BaseMeasurement):
+    def __init__(self, id, url):
+        super(YouTubeMeasurement, self).__init__(id=id)
+        validated_url = validators.url(url)
+        if isinstance(validated_url, ValidationFailure):
+            raise ValueError("`{url}` is not a valid url".format(url=url))
+        self.id = id
+        self.url = url
+        self.progress_dicts = []
+
+    def measure(self):
+        return self._get_youtube_result(self.url)
+
+    def _get_youtube_result(self, url):
+        # Unique filename from process ID and timestamp
+        file_dir = "{}/youtube-dl_{}".format(tempfile.gettempdir(), os.getpid())
+        filename = "{}/youtube-dl_{}/{}_%(title)s.%(ext)s".format(
+            tempfile.gettempdir(), os.getpid(), int(time.time())
+        )
```

Comment:
I hit an error in testing (see below). I thought it might be related to the spaces in the `title` part of the filename but looking closer I see that it's the `.f140` missing from the name of the file saved on disk.

I'm not sure if it's a good idea to use the `title` in the temporary filename at all as you never know what you're gonna get. If you absolutely must use the title you should at least normalise it so that it only includes lower-case characters and underscores or something along those lines.

```
In [1]: import uuid                                                                                                                                                                                                

In [2]: from measurement.plugins.youtube.measurements import YouTubeMeasurement                                                                                                                                    

In [3]: ytm = YouTubeMeasurement( 
   ...:     "{uuid}".format(uuid=uuid.uuid4()), 
   ...:     "https://www.youtube.com/watch?v=9F5RoLF9_RU" 
   ...:     )                                                                                                                                                                                                      

In [4]: ytm.measure()                                                                                                                                                                                              
Out[4]: YouTubeMeasurementResult(id='e0c78fd5-fa08-44ad-97fd-5372dda6f49f', errors=[Error(key='youtube-file', description='Could not remove file!!', traceback="[Errno 2] No such file or directory: '/tmp/youtube-dl_741866/1596754637_Shockwave from bomb.f140.m4a'")], download_rate=None, download_rate_unit=None, download_size=None, download_size_unit=None, url='https://www.youtube.com/watch?v=9F5RoLF9_RU', elapsed_time=None, elapsed_time_unit=None)

In [5]: ll /tmp/youtube-dl_741866/                                                                                                                                                                                 
total 408
-rw-rw-r-- 1 james 414009 Jul 21 00:33 '1596754637_Shockwave from bomb.mp4'
```

Reply:
Good point, not really any value to including the title in there anyway.

Reply:
Did you test this after the update? It's still not working for me (see below).

The file path that it's attempting to delete is different to what's on disk: `/tmp/youtube-dl_906096/1597133901.f140` vs `/tmp/youtube-dl_906096/1597133901.mkv` (I mentioned the `.f140` in my last comment).

```
In [1]: import uuid

In [2]: from measurement.plugins.youtube.measurements import YouTubeMeasurement

In [3]: ytm = YouTubeMeasurement(
   ...:     "{uuid}".format(uuid=uuid.uuid4()),
   ...:     "https://www.youtube.com/watch?v=9F5RoLF9_RU"
   ...:     )

In [4]: from measurement.plugins.youtube.measurements import YouTubeMeasurement

In [5]: ytm.measure()
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
Out[5]: YouTubeMeasurementResult(id='1576439b-ce46-4fb7-8ea2-6c4346c8e3d2', errors=[Error(key='youtube-file', description='Could not remove file!!', traceback="[Errno 2] No such file or directory: '/tmp/youtube-dl_906096/1597133901.f140'")], download_rate=None, download_rate_unit=None, download_size=None, download_size_unit=None, url='https://www.youtube.com/watch?v=9F5RoLF9_RU', elapsed_time=None, elapsed_time_unit=None)
```

```
$ ll /tmp/youtube-dl_906096/
total 708
drwxrwxr-x  2 james james   4096 Aug 11 18:18 .
drwxrwxrwt 36 root  root  282624 Aug 11 18:18 ..
-rw-rw-r--  1 james james 431821 Aug  9 13:43 1597133901.mkv
```

Reply:
@amorphic I've changed it such that it just dumps the whole directory now, are you still running into errors?

Reply:
@jamesaj23 yep working now!

```
In [1]: import uuid

In [2]: from measurement.plugins.youtube.measurements import YouTubeMeasurement

In [3]: ytm = YouTubeMeasurement(
   ...:     "{uuid}".format(uuid=uuid.uuid4()),
   ...:     "https://www.youtube.com/watch?v=9F5RoLF9_RU"
   ...:     )

In [4]: ytm.measure()
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
Out[4]: YouTubeMeasurementResult(id='a0b204df-992d-453f-b9bc-ca0bec73fda1', errors=[], download_rate=24236468.738593042, download_rate_unit=<NetworkUnit.bit_per_second: 'bit/s'>, download_size=245734, download_size_unit=<StorageUnit.bytes: 'B'>, url='https://www.youtube.com/watch?v=9F5RoLF9_RU', elapsed_time=0.1595897674560547, elapsed_time_unit=<TimeUnit.second: 's'>)
```

:-)